openstef/model/metamodels/grouped_regressor.py

```diff
@@ -0,0 +1,141 @@
+# SPDX-FileCopyrightText: 2017-2021 Alliander N.V. <korte.termijn.prognoses@alliander.com> # noqa E501>
+#
+# SPDX-License-Identifier: MPL-2.0
+import joblib
+import pandas as pd
+from sklearn.base import BaseEstimator, RegressorMixin, MetaEstimatorMixin, clone
+from sklearn.utils.validation import check_is_fitted
+
+
+class GroupedRegressor(BaseEstimator, RegressorMixin, MetaEstimatorMixin):
+    """Meta-model that trains an instance of the base estimator for each key of a groupby operation applied on the data.
+    The base estimator is a sklearn regressor, the groupby is performed on the columns specified in parameters.
+    Moreover fit and predict methods can be performed in parallel for each group key thanks to joblib.
+    Example:
+
+    data = | index | group | x0 | x1 | x3 | y |
+           |   0   |   1   | .. | .. | .. | . |
+           |   1   |   2   | .. | .. | .. | . |
+           |   2   |   1   | .. | .. | .. | . |
+           |   3   |   2   | .. | .. | .. | . |
+
+           [              X              ][ Y ]
+
+    The GroupedRegressor on the data with the group_columns='group' fits 2 models :
+        the model 1 with the row 0 and 2, columns x0, x1 and x3 as the features and column y as the target.
+        the model 2 with the row 1 and 3, columns x0, x1 and x3 as the features and column y as the target.
+
+    Parameters
+    ----------
+    base_estimator: RegressorMixin
+        Regressor .
+
+    group_columns : str, int, list
+        Name(s) of the column(s) used as the key for groupby operation.
+
+    n_jobs : int default=1
+        The maximum number of concurrently running jobs,
+         such as the number of Python worker processes when backend=”multiprocessing”
+         or the size of the thread-pool when backend=”threading
+
+     Attributes
+    ----------
+        feature_names_: list(str)
+            All input feature (without group_columns).
+
+        estimators_: dict(str, RegressorMixin)
+            Dictionnary that stocks fitted estimators for each group.
+            The keys are the keys of grouping and the values are the regressors fitted on the grouped data.
+    """
+
+    def __init__(self, base_estimator, group_columns, n_jobs=1):
```

Comment:
A general remark: I see the methods do not have typehints (both input and output), could you please add these to all mehtods?